### ISSUE TITLE
Multiple handlers to one adapter seem to deliver corrupt r.AdapterConfig.Value (non deterministic)

### ISSUE THREAD

**Issue by johnholiver** [istio/istio#9560]
**Describe the bug**
r.AdapterConfig.Value carries a configuration set at a handler into an adapter.
If the adapter can be accessed from multiple handlers, it seems that the AdapterConfig value sent to the adapter from the Mixer can get corrupted, mixing values from different handlers.
Corrupt AdapterConfig values fail to be unmarshalled on the adapter, resulting in different errors such as: illegal wireType 6, illegal wireType 7 or unexpected EOF. In the case of failure it's not unusual to print the r.AdapterConfig.Value as string and to find partial content from an incorrect handler. 

_Non-determinism_
As mentioned above errors while unmarshalling AdapterConfig can vary.
However, it also works eventually! Removing and adding again handlers, restarting mixer and the adapter sometimes make the setup work.
Additionally, having just one handler per adapter at a time seems to work perfectly. So, in my case I had 3 handlers, and I could see that each of them was correctly configured/working by keeping only one of the handlers CRD up at a time.

**Expected behavior**
One can deploys N handlers to the same adapter and, each time a specific handler is used, the adapter will receive the correct AdapterConfig, which has been writtent on that specific handler.

**Steps to reproduce the bug**
As a non-deterministic issue, reproducing it might be difficult.
I'll provide an example to ilustrate. Bare in mind that those are not exact copies from my own CRDs as I generate them using Helm. I also trimmed unimportant stuff.
Below we find 2 handlers named after the microservices which they are responsible to check requests from: "entrance" and "tokend". We also can find one of the rules to ilustrate how these handlers are reached.
```yaml
apiVersion: config.istio.io/v1alpha2
kind: handler
metadata:
  name: ditto-entrance
  namespace: istio-system
spec:
  adapter: ditto
  connection:
    address: dittomixeradapter:90
  params:
    api_requirements:
    - endpoints:
      - endpoint: /endpoint
        method: GET
        name: endpoint
        requirements:
        - actions:
          - get
          resource: endpoint
      version: null
```
```yaml
apiVersion: config.istio.io/v1alpha2
kind: handler
metadata:
  name: ditto-tokend
  namespace: istio-system
spec:
  adapter: ditto
  connection:
    address: dittomixeradapter:90
  params:
    api_requirements:
    - endpoints:
      - endpoint: /resourcepolicy
        method: POST
        name: resource-policy-create
        requirements:
        - actions:
          - create
          resource: resource-policy
      - endpoint: /resourcepolicy
        method: DELETE
        name: resource-policy-delete
        requirements:
        - actions:
          - delete
          resource: resource-policy
      - endpoint: /resourcepolicy
        method: GET
        name: resource-policy-query
        requirements:
        - actions:
          - query
          resource: resource-policy
      version: /api/v1
```
```yaml
apiVersion: "config.istio.io/v1alpha2"
kind: rule
metadata:
  name: entrance
spec:
  match: match(destination.service.name, "entrance")
  actions:
  - handler: ditto-entrance.handler.istio-system
    instances:
    - ditto-authorization.instance.istio-system
```

As one can observe, the _entrance_ handler has references on params to strings such as _endpoint_ while the _tokend_ handler has references to _resourcepolicy_. The beginning of my adapter HandleAuthorization method looks like this:
```go
func (s *Adapter) HandleAuthorization(context context.Context, r *authorization.HandleAuthorizationRequest) (*v1beta1.CheckResult, error) {
   service := r.Instance.Action.Service
   signedResourcePolicy := r.Instance.Subject.Properties["signedResourcePolicy"].GetStringValue()
   protocol := r.Instance.Action.Properties["protocol"].GetStringValue()
   path := r.Instance.Action.Path
   method := strings.ToUpper(r.Instance.Action.Method)

   fmt.Printf("Received Authorization Request:\n"+
      "Service: %v\n"+
      "SignedResourcePolicy: %v\n"+
      "Protocol: %v\n"+
      "Path: %v\n"+
      "Method: %v\n",
      service, signedResourcePolicy, protocol, path, method)  
   cfg := &config.Params{}
   fmt.Printf("AdapterConfig: %v \n", string(r.AdapterConfig.Value))
   if r.AdapterConfig != nil {
      if err := cfg.Unmarshal(r.AdapterConfig.Value); err != nil {
         log.Errorf("error unmarshalling adapter config: %v", err)
         return nil, err
      }
   }
   ...
}
```
The occourance of fmt.Printf helps us to see the issue. And the log for one of the corrupted Authorization Request received is:
```
2018/10/25 15:46:33 stdout: Received Authorization Request:
2018/10/25 15:46:33 stdout: Service: test-dittomixeradapter-entrance
2018/10/25 15:46:33 stdout: SignedResourcePolicy: eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE1NDA0ODI2OTMsInN1YiI6IkpvaG4iLCJyZXNvdXJjZVBvbGljaWVzIjpbeyJpZCI6IjViZDFlMjg1YTcwZjE1YjJhNTEzMDE0OCIsInVzZXIiOiJKb2huIiwic3RhdGVtZW50Ijp7ImFjdGlvbiI6WyJnZXQiXSwicmVzb3VyY2UiOiJlbmRwb2ludCJ9fV19.A1snqKR8sul-lWzxkp_rTchuNmHFmVDBFmmpIPBoff3d3N9UfLyRpib6UrApMiGiJDlrn9p3u_ejXVQLQnjhtdy83NXXJ53hUtbPHQUU8gaVPS2STL10CILlUmQy5xIY9sagABMG608Qjp18HJJPQ5blSpA18p5ZB8dwy_DFDEqKXB5-FmC6FJyi1lviTdEnU4sEIpc7bK9Lof5KhOx9Wx-YwrrMNOPMDyWyjKB59lJOVZoi8tNtaTFP3EElPpseaTtM_JZxVQ9pn6Kn58h7BssER69KMyIt9dDLYNEfks_8xDE1scYr70TYg7bp4cZ53t4uoPAZ-LdQuL1FeqBmeA
2018/10/25 15:46:33 stdout: Protocol: http
2018/10/25 15:46:33 stdout: Path: /endpoint
2018/10/25 15:46:33 stdout: Method: GET
2018/10/25 15:46:33 stdout: AdapterConfig: /resourcepolicyPUT
2018/10/25 15:46:33 stdout: resource-policy-update" 
2018/10/25 15:46:33 stdout: 2018-10-25T15:46:33.574214Z	error	error unmarshalling adapter config: unexpected EOF
```
The main point here is that the service that I was forwarding my request is _entrance_, yet, while printing AdapterConfig there are references to _resource-policy_, which are only set on a different handler configuration (_tokend_ one).

Unfortunately, the mixer logs with full debug mode on does not bring light to the exact AdapterConfig value which is sent the adapter. Also, investigating ahead on the construction of the handlers and routing tables on the mixer logs didn't lead to many clues:

```
[Routing ExpectedTable]
ID: 8
[#0] TEMPLATE_VARIETY_CHECK {V}
  [#0] default {NS}
    [#0] ditto-entrance.handler.istio-system {H}
      [#0]
        Condition: match(destination.service.name, "test-dittomixeradapter-entrance")
        [#0] ditto-authorization.instance.istio-system {I}
    [#1] ditto-exit.handler.istio-system {H}
      [#0]
        Condition: match(destination.service.name, "test-dittomixeradapter-exit")
        [#0] ditto-authorization.instance.istio-system {I}
    [#2] ditto-tokend.handler.istio-system {H}
      [#0]
        Condition: match(destination.service.name, "test-dittomixeradapter-tokend")
        [#0] ditto-authorization.instance.istio-system {I}
```
```
2018-10-25T13:31:45.721770Z	debug	Processing incoming handler config: name='ditto-tokend.handler.istio-system'
adapter:"ditto" params:map[api_requirements:[map[endpoints:[map[endpoint:/healthz method:GET name:healthz requirements:[]]] version:<nil>] map[endpoints:[map[endpoint:/sign method:POST name:resource-policy-sign requirements:[]] map[requirements:[map[resource:resource-policy actions:[create]]] endpoint:/resourcepolicy method:POST name:resource-policy-create] map[endpoint:/resourcepolicy method:DELETE name:resource-policy-delete requirements:[map[actions:[delete] resource:resource-policy]]] map[endpoint:/resourcepolicy method:GET name:resource-policy-query requirements:[map[actions:[query] resource:resource-policy]]] map[endpoint:/resourcepolicy method:PUT name:resource-policy-update requirements:[map[actions:[update] resource:resource-policy]]]] version:/api/v1]]] connection:<address:"dev-dittomixeradapter:90" > 
--- 
2018-10-25T13:31:45.726634Z	debug	Processing incoming handler config: name='ditto-exit.handler.istio-system'
adapter:"ditto" params:map[api_requirements:[map[endpoints:[map[requirements:[map[actions:[get] resource:endpoint]] endpoint:/Endpoint name:grpc-endpoint]] version:/pb.ExitService] map[endpoints:[map[name:endpoint requirements:[map[actions:[get] resource:endpoint]] endpoint:/endpoint method:GET]] version:<nil>]]] connection:<address:"dev-dittomixeradapter:90" > 
--- 
2018-10-25T13:31:45.727818Z	debug	Processing incoming handler config: name='ditto-entrance.handler.istio-system'
adapter:"ditto" params:map[api_requirements:[map[endpoints:[map[endpoint:/endpoint method:GET name:endpoint requirements:[map[actions:[get] resource:endpoint]]]] version:<nil>]]] connection:<address:"dev-dittomixeradapter:90" > 
```

**Version**
Istio: v1.0.0
K8s: v1.11.1

**Installation**
Deployed thru Helm

**Environment**
Minikube

_This bug has been originally discussed over the [Istio Policy & Telemetry](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/istio-policies-and-telemetry/APeIpGRZOXg) google group_

**Comment by hklai** [comment]
@kuat 1.1 or 1.2?

**Comment by mandarjog** [comment]
1.1

**Comment by mandarjog** [comment]
@kyessenov any progress on this ?

**Comment by kyessenov** [comment]
No luck. @johnholiver we need more details to diagnose this further.
I tried to simulate your conditions with generating a constant load to mixer, and then trigger three handlers to the same adapter. I cannot get the config unmarshal to produce an error. I have tried to restarting the adapter and mixer to trigger some edge cases, but the proto-encoded bytes seem to be consistently OK.

I suspect the error is somewhat transient in the set-up or this has something to do with your particular set-up. 

Here are some details on my set-up:
- I'm generating load with:
```
go run mixer/cmd/mixc/main.go check -r 10000000 -c 12 > /dev/null
```
- Added a dummy ditto template / adapter:
```proto
syntax = "proto3";

package ditto;

import "mixer/adapter/model/v1beta1/extensions.proto";

option (istio.mixer.adapter.model.v1beta1.template_variety) = TEMPLATE_VARIETY_CHECK;

message Template {
  Action action = 1;
}

message Action {
  string service = 1;
  map<string,string> properties = 2;
  string path = 3;
  string method = 4;
}
```

```proto
package config;

message Requirement {
  repeated string actions = 1;
  string resource = 2;
}

message Endpoint {
  string endpoint = 1;
  string method = 2;
  string name = 3;
  repeated Requirement requirements = 4;
}

message APIRequirement{
  repeated Endpoint endpoints = 1;
  string version = 2;
}

message Params{
  repeated APIRequirement api_requirements = 1;
}
```

```go
type adapter struct{}

func (adapter) HandleDitto(ctx context.Context, req *ditto.HandleDittoRequest) (*v1beta1.CheckResult, error) {
        cfg := &config.Params{}
        if err := cfg.Unmarshal(req.AdapterConfig.Value); err != nil || cfg.ApiRequirements[0].Endpoints[0].Method == "" {
                log.Printf("error unmarshalling adapter config: %v %#v", err, cfg)
                return nil, err
        }

        return &v1beta1.CheckResult{}, nil
}

func main() {
        listener, err := net.Listen("tcp", ":9090")
        if err != nil {
                panic(err)
        }
        server := grpc.NewServer()
        ditto.RegisterHandleDittoServiceServer(server, adapter{})
        server.Serve(listener)
}
```

- Added rules / handlers:
```yaml
apiVersion: config.istio.io/v1alpha2
kind: instance
metadata:
  name: ditto-authorization
  namespace: istio-system
spec:
  template: ditto
  params:
    action:
      path: '"test"'
---
apiVersion: "config.istio.io/v1alpha2"
kind: rule
metadata:
  name: entrance
  namespace: istio-system
spec:
  actions:
  - handler: ditto-entrance.handler.istio-system
    instances:
    - ditto-authorization.instance.istio-system
---
apiVersion: "config.istio.io/v1alpha2"
kind: rule
metadata:
  name: exit
  namespace: istio-system
spec:
  actions:
  - handler: ditto-exit.handler.istio-system
    instances:
    - ditto-authorization.instance.istio-system
---
apiVersion: "config.istio.io/v1alpha2"
kind: rule
metadata:
  name: tokend
  namespace: istio-system
spec:
  actions:
  - handler: ditto-tokend.handler.istio-system
    instances:
    - ditto-authorization.instance.istio-system
```

```yaml
apiVersion: config.istio.io/v1alpha2
kind: handler
metadata:
  name: ditto-entrance
  namespace: istio-system
spec:
  adapter: ditto
  connection:
    address: ":9090"
  params:
    api_requirements:
    - endpoints:
      - endpoint: /entrance
        method: GET
        name: endpoint
        requirements:
        - actions:
          - get
          resource: endpoint
      version: null
---
apiVersion: config.istio.io/v1alpha2
kind: handler
metadata:
  name: ditto-exit
  namespace: istio-system
spec:
  adapter: ditto
  connection:
    address: ":9090"
  params:
    api_requirements:
    - endpoints:
      - endpoint: /exit
        method: GET
        name: endpoint
        requirements:
        - actions:
          - get
          resource: endpoint
      version: null
---
apiVersion: config.istio.io/v1alpha2
kind: handler
metadata:
  name: ditto-tokend
  namespace: istio-system
spec:
  adapter: ditto
  connection:
    address: ":9090"
  params:
    api_requirements:
    - endpoints:
      - endpoint: /resourcepolicy
        method: POST
        name: resource-policy-create
        requirements:
        - actions:
          - create
          resource: resource-policy
      - endpoint: /resourcepolicy
        method: DELETE
        name: resource-policy-delete
        requirements:
        - actions:
          - delete
          resource: resource-policy
      - endpoint: /resourcepolicy
        method: GET
        name: resource-policy-query
        requirements:
        - actions:
          - query
          resource: resource-policy
      version: /api/v1
```

- Added a dummy script to trigger runtime changes:
```sh
cat <<EOF | kubectl create -f -
apiVersion: config.istio.io/v1alpha2
kind: handler
metadata:
  name: test
  namespace: istio-system
spec:
  adapter: ditto
  connection:
    address: ":9090"
  params:
    api_requirements:
    - endpoints:
      - endpoint: /test
        method: GET
        name: endpoint
        requirements:
        - actions:
          - get
          resource: endpoint
      version: null
EOF
cat <<EOF | kubectl create -f -
apiVersion: "config.istio.io/v1alpha2"
kind: rule
metadata:
  name: test
  namespace: istio-system
spec:
  actions:
  - handler: test.handler.istio-system
    instances:
    - ditto-authorization.instance.istio-system

EOF
sleep 2
kubectl delete rule test -n istio-system
kubectl delete handler test -n istio-system
```

**Comment by johnholiver** [comment]
@kyessenov I started to copy paste all the code you put here to setup the same simulation locally, but in case you read before I finish, do you have a zipped file with all this that I can reuse?
Also, which Istio version/commit-sha did you test against?

My plan is to try your simulation, retry my own setup and if mine is broken and yours working I'll start to copy your simulation into my environment and mine to yours until I get an AHA moment.

**Comment by kyessenov** [comment]
@johnholiver Sorry, I should have shared my branch immediately:
https://github.com/kyessenov/istio/tree/nondet

I have started off the tip of release 1.1. I don't think we made big changes in the underlying proto/gRPC code to matter.

I think most details are already up above. I pushed the template and the adapter config first:
```sh
kubectl apply -f ditto/tmpl.yaml; kubectl apply -f ditto/config/ditto.yaml
```

I started mixer server with this command:
```sh
go run mixer/cmd/mixs/main.go server --numCheckCacheEntries=0 --configStoreURL="k8s:///$HOME/.kube/config"
```

Let me know if there is something I can do to help out to get to the bottom of this.
